--- a/src/providers/data_provider_be.c
+++ b/src/providers/data_provider_be.c
@@ -553,6 +553,7 @@
     char *srv_name = NULL;
     struct main_context *main_ctx;
     char *confdb_path;
+    int init_as_user;
     int ret;
     uid_t uid;
     gid_t gid;
@@ -564,6 +565,8 @@
         SSSD_SERVER_OPTS(uid, gid)
         {"domain", 0, POPT_ARG_STRING, &be_domain, 0,
          _("Domain of the information provider (mandatory)"), NULL },
+        {"init-as-user", ' ', POPT_ARG_NONE, &init_as_user, 0,
+         _("Switch user before initialisation"), NULL },
         POPT_TABLEEND
     };
 
@@ -603,7 +606,10 @@
     confdb_path = talloc_asprintf(NULL, CONFDB_DOMAIN_PATH_TMPL, be_domain);
     if (!confdb_path) return 2;
 
-    ret = server_setup(srv_name, 0, 0, 0, confdb_path, &main_ctx);
+    ret = server_setup(srv_name, 0,
+                       init_as_user ? uid : 0,
+                       init_as_user ? gid : 0,
+                       confdb_path, &main_ctx);
     if (ret != EOK) {
         DEBUG(SSSDBG_FATAL_FAILURE, "Could not set up mainloop [%d]\n", ret);
         return 2;
@@ -631,17 +637,19 @@
         return 3;
     }
 
-    ret = chown_debug_file(NULL, uid, gid);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_MINOR_FAILURE,
-              "Cannot chown the debug files, debugging might not work!\n");
-    }
+    if (!init_as_user) {
+        ret = chown_debug_file(NULL, uid, gid);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_MINOR_FAILURE,
+                  "Cannot chown the debug files, debugging might not work!\n");
+        }
 
-    ret = become_user(uid, gid);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_FUNC_DATA,
-              "Cannot become user [%"SPRIuid"][%"SPRIgid"].\n", uid, gid);
-        return ret;
+        ret = become_user(uid, gid);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_FUNC_DATA,
+                  "Cannot become user [%"SPRIuid"][%"SPRIgid"].\n", uid, gid);
+            return ret;
+        }
     }
 
     DEBUG(SSSDBG_TRACE_FUNC, "Backend provider (%s) started!\n", be_domain);
