--- a/src/providers/krb5/krb5_child.c
+++ b/src/providers/krb5/krb5_child.c
@@ -1162,7 +1162,7 @@ static errno_t changepw_child(struct krb
     kerr = krb5_get_init_creds_password(kr->ctx, kr->creds, kr->princ,
                                         discard_const(password),
                                         prompter, kr, 0,
-                                        SSSD_KRB5_CHANGEPW_PRINCIPAL,
+                                        discard_const(SSSD_KRB5_CHANGEPW_PRINCIPAL),
                                         kr->options);
     DEBUG(SSSDBG_TRACE_INTERNAL,
           "chpass is%s using OTP\n", kr->otp ? "" : " not");
@@ -1344,7 +1344,7 @@ static errno_t tgt_req_child(struct krb5
     kerr = krb5_get_init_creds_password(kr->ctx, kr->creds, kr->princ,
                                         discard_const(password),
                                         sss_krb5_prompter, kr, 0,
-                                        SSSD_KRB5_CHANGEPW_PRINCIPAL,
+                                        discard_const(SSSD_KRB5_CHANGEPW_PRINCIPAL),
                                         kr->options);
 
     krb5_free_cred_contents(kr->ctx, kr->creds);
