Index: pkg-sssd/src/config/SSSDConfig.py
===================================================================
--- pkg-sssd.orig/src/config/SSSDConfig.py	2011-11-22 20:26:08.782749600 +0200
+++ pkg-sssd/src/config/SSSDConfig.py	2011-11-22 20:26:27.879317350 +0200
@@ -254,9 +254,9 @@
         SSSDChangeConf.__init__(self)
         #TODO: get these from a global setting
         if not schemafile:
-            schemafile = '/etc/sssd/sssd.api.conf'
+            schemafile = '/usr/share/sssd/sssd.api.conf'
         if not schemaplugindir:
-            schemaplugindir = '/etc/sssd/sssd.api.d'
+            schemaplugindir = '/usr/share/sssd/sssd.api.d'
 
         try:
             #Read the primary config file
@@ -1155,10 +1155,10 @@
 
         schemafile:
           The path to the api schema config file. Usually
-          /etc/sssd/sssd.api.conf
+          /usr/share/sssd/sssd.api.conf
         schemaplugindir:
           The path the directory containing the provider schema config files.
-          Usually /etc/sssd/sssd.api.d
+          Usually /usr/share/sssd/sssd.api.d
 
         === Returns ===
         The newly-created SSSDConfig object.
