diff --git a/src/providers/tacplus/tacplus_common.c b/src/providers/tacplus/tacplus_common.c
index 72688ff54939..e91830c92d99 100644
--- a/src/providers/tacplus/tacplus_common.c
+++ b/src/providers/tacplus/tacplus_common.c
@@ -264,7 +264,7 @@ int tacplus_author(const char *user, const char *service, const char *proto,
                 if (strlen(key) > 1)
                     key_nosep = strndup(key, strlen(key)-1);
 
-                if strchr(key, '=') {
+                if (strchr(key, '=')) {
                     sss_log(SSS_LOG_WARNING, "Unsupported mandatory TACACS+ attribute '%s'",
                             key_nosep ? key_nosep : key);
                     status = TAC_PLUS_AUTHOR_STATUS_FAIL;
-- 
2.20.1

