From: Dominik George <natureshadow@debian.org>
Subject: Allow building authdata plugin with krb5 1.17
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=923125
--- a/configure
+++ b/configure
--- a/src/external/pac_responder.m4
+++ b/src/external/pac_responder.m4
@@ -18,7 +18,8 @@ then
         Kerberos\ 5\ release\ 1.13* | \
         Kerberos\ 5\ release\ 1.14* | \
         Kerberos\ 5\ release\ 1.15* | \
-        Kerberos\ 5\ release\ 1.16*)
+        Kerberos\ 5\ release\ 1.16* | \
+        Kerberos\ 5\ release\ 1.17*)
             krb5_version_ok=yes
             AC_MSG_RESULT([yes])
             ;;
