diff --git a/src/external/systemd.m4 b/src/external/systemd.m4
index dbced0d..92b9c99 100644
--- a/src/external/systemd.m4
+++ b/src/external/systemd.m4
@@ -8,16 +8,21 @@ AC_DEFUN([AM_CHECK_SYSTEMD],
 ])
 
 AM_COND_IF([HAVE_SYSTEMD],
-           [PKG_CHECK_MODULES([SYSTEMD_LOGIN], [libsystemd-login],
-            [AC_DEFINE_UNQUOTED(HAVE_SYSTEMD_LOGIN, 1, [Build with libsystemdlogin support])],
-            [AC_MSG_NOTICE([Build without libsystemd-login support])])])
+           [PKG_CHECK_MODULES([SYSTEMD_LOGIN], [libsystemd],
+            [AC_DEFINE_UNQUOTED(HAVE_SYSTEMD_LOGIN, 1, [Build with libsystemd login support])],
+            [PKG_CHECK_MODULES([SYSTEMD_LOGIN], [libsystemd-login],
+             [AC_DEFINE_UNQUOTED(HAVE_SYSTEMD_LOGIN, 1, [Build with libsystemd login support])],
+             [AC_MSG_NOTICE([Build without libsystemd-login support])])])])
 
 dnl A macro to check presence of journald on the system
 AC_DEFUN([AM_CHECK_JOURNALD],
 [
-       PKG_CHECK_MODULES(JOURNALD,
-                         libsystemd-journal,
-                         [AC_DEFINE_UNQUOTED([WITH_JOURNALD], 1, [journald is available])])
+       PKG_CHECK_MODULES([JOURNALD],
+                         [libsystemd],
+                         [AC_DEFINE_UNQUOTED([WITH_JOURNALD], 1, [journald is available])]
+                         [PKG_CHECK_MODULES([JOURNALD],
+                                            [libsystemd-journal],
+                                            [AC_DEFINE_UNQUOTED([WITH_JOURNALD], 1, [journald is available])])])
        dnl Some older versions of pkg-config might not set these automatically
        dnl while setting CFLAGS and LIBS manually twice doesn't hurt.
        AC_SUBST([JOURNALD_CFLAGS])
