--- a/src/providers/tacplus/tacplus_common.c
+++ b/src/providers/tacplus/tacplus_common.c
@@ -131,6 +131,8 @@
                 }
             }
             else if (strcmp("local-user-name=", key) == 0) {
+                sss_log(DANOS_LOCAL_USER_LOG_LVL, "Support for the 'local-user-name' TACACS+ "
+                        "attribute is deprecated and will be removed in a future release");
                 dbus_message_iter_next(&dictEntryIter);
                 if (local_account) {
                     dbus_message_iter_get_basic(&dictEntryIter, &tmp);
