diff -urNad a/server/examples/sssd.conf b/server/examples/sssd.conf
--- a/src/examples/sssd.conf	2010-01-18 13:09:10.000000000 -0500
+++ b/src/examples/sssd.conf	2010-01-18 13:18:18.000000000 -0500
@@ -13,7 +13,7 @@
 # Add new domain configurations as [domain/<NAME>] sections, and
 # then add the list of domains (in the order you want them to be
 # queried) to the "domains" attribute below and uncomment it.
-; domains = LOCAL,LDAP
+domains = LOCAL
 
 [nss]
 # The following prevents SSSD from searching for the root user/group in
@@ -45,12 +45,12 @@
 # Example LOCAL domain that stores all users natively in the SSSD internal
 # directory. These local users and groups are not visible in /etc/passwd; it
 # now contains only root and system accounts.
-; [domain/LOCAL]
-; description = LOCAL Users domain
-; id_provider = local
-; enumerate = true
-; min_id = 500
-; max_id = 999
+[domain/LOCAL]
+description = LOCAL Users domain
+id_provider = local
+enumerate = true
+min_id = 500
+max_id = 999
 
 # Example native LDAP domain
 # ldap_schema can be set to "rfc2307", which uses the "memberuid" attribute
